File: two_pointers/259. 3Sum Smaller/three_sum_smaller.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer}
def three_sum_smaller(nums, target)
  nums = nums.sort
  n = nums.length
  return 0 if n == 0
  count = 0
  for i in (0..n-2)
  	count += search_pair(nums, target, i)
  end
  return count
end


def search_pair(nums, target, i)
	left = i + 1 
	right = nums.length - 1
	count = 0
	while left < right		
		sum = nums[i] + nums[left] + nums[right]
		if sum < target
			#we can get a diff right - left. If right is works for this process , between the right and left indices also works in a sorted arr.
			count += right - left  
			left += 1
		else
			right -= 1
		end
	end
	return count
end


nums = [-1, 0, 2, 3]
target = 3

res = three_sum_smaller(nums, target)
puts "res : #{res}"

File: two_pointers/259. 3Sum Smaller/threeSumSmaller.py
def triplet_with_smaller_sum(arr, target):
  arr.sort()
  count = 0
  #we can take len(arr) - 2 , because we have 2 pointer left and right. 
  for i in range(len(arr)-2):
    count += search_pair(arr, target, i)
  return count


def search_pair(arr, target, i):
  count = 0
  left, right = i + 1, len(arr) - 1
  while (left < right):
    if  arr[i] + arr[left] + arr[right] < target:  # found the triplet
      count += right - left
      left += 1
    else:
      right -= 1  # we need a pair with a smaller sum
  return count


def main():
  # print(triplet_with_smaller_sum([-1, 0, 2, 3], 3))
  # print(triplet_with_smaller_sum([-1, 4, 2, 1, 3], 5))
  print(triplet_with_smaller_sum([-2,0,1,3], 2))

main()

